### DISCUSSION TITLE
How are the changelogs generated?

### DISCUSSION THREAD

**Discussion by pastelmind** [typescript-eslint/typescript-eslint#6626] [Technical Discussions]
Hi. I'm working on a monorepo project and would like to learn from bigger projects like typescript-eslint. I'd like to know how the maintainers generate changelogs for each package, and post them to GitHub. Is the process automated? If so, how?

I know of tools that extract changelogs from commit messages, such as [release-it](https://github.com/release-it/release-it). However, AFAIK typescript-eslint doesn't appear to use such a package. I examined the GitHub workflows and `package.json` of subpackages, but I can't seem to figure out how everything works.

It would be helpful if you could point me to a script or file that drives the changelog generation process. Thanks!

**Comment by bradzacher** [comment]
We enforce PR titles match semantic commit formatting using [this GH action workflow](https://github.com/typescript-eslint/typescript-eslint/blob/main/.github/workflows/semantic-pr-titles.yml), and we use the PR title as the commit message when we squash-merge the PR:
![image](https://user-images.githubusercontent.com/7462525/224925639-4f8f054d-c416-48ff-8e33-f3c62dc7081a.png)

Our release script is in our private repo for various reasons, but TL;DR - we use lerna for this
```
# Bump the package versions according to the commit data, create a commit for the release,
# and push it all to the typescript-eslint/typescript-eslint repo,
# including creating a new entry on GitHub Releases with the updated CHANGELOG
# (--yes will skip the confirmation prompt)
GH_TOKEN=foo npx lerna version --loglevel=silly --yes --conventional-commits --exact --force-publish --github-release -m "chore: publish %s"
```

Lerna will scan the commit messages since the last release and automatically generate the release notes (as well as appropriately bump the version)

**Reply by pastelmind** [reply]
Wow, that was fast. Thank you.

If it doesn't bother you, I'd like to know how you came to choose Lerna. Was it because Lerna provided features that other solutions did not, or perhaps because your team was already familiar with it?

**Reply by bradzacher** [reply]
We chose lerna because this repo is 5 years old - so at the time lerna was the default choice for monorepo setups!

Other parts of our tool chain have changed over the years (eg we use nx and yarn workspaces for our commands and linking), but lerna still manages our release because it continues to do exactly what we need it to do!

As the saying goes "if it ain't broke, don't fix it" 😀
